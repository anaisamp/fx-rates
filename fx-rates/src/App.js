import React, { Component } from 'react';
import { connect } from 'react-redux';
import './App.css';

import { getExchangeRate } from './actions/getExchangeRate';


const currencies = [
  { symbol: 'GBP', currency: '£' },
  { symbol: 'EUR', currency: '€' },
  { symbol: 'USD', currency: '$' }
];

class App extends Component {
  constructor(props){
    super(props);
    this.state = {
      from: 'EUR',
      to: 'GBP',
      toValue: '',
      fromValue: '',
    }
  }

  componentDidMount() {
    this.props.getExchangeRate(this.state.from);
    this.props.getExchangeRate(this.state.to);
    //this.interval = setInterval(() => this.props.getExchangeRate(), 10000); // service worker??
  }

  componentWillUnmount() {
    clearInterval(this.interval);
  }

  shouldComponentUpdate(nextProps, nextState) {
   return this.props !== nextProps || this.state !== nextState;
  }

  handleExchangeRate = (e) => {
    // send selected rates as props, then filter !currency
    const value = e.target.value;
    const currencyBase = e.target.name;
    const currencyTo = currencyBase === this.state.from ? this.state.to : this.state.from;
    const inputTo = currencyBase === this.state.from ? 'toValue' : 'fromValue';
    const inputFrom = currencyBase === this.state.from ? 'fromValue' : 'toValue';

    const exhangeValue = parseFloat(value) * this.props[currencyBase][currencyTo];

    this.setState({
      [inputFrom]: parseFloat(value),
      [inputTo]: exhangeValue.toFixed(2)
    });

  }

  handleCurrencyFromChange = (e) => {
    const toV = this.state.to;
    const oldFrom = this.state.from;
    const updatedFrom = e.target.value;

    if (toV === updatedFrom) {
      this.setState({
        to: oldFrom,
        from: updatedFrom,
        fromValue: '',
        toValue: '',
      });
    }
    else {
      this.setState({
        [e.target.name]: e.target.value,
        fromValue: '',
        toValue: '',
      });
    }
    this.props.getExchangeRate(updatedFrom);
  }

  handleCurrencyToChange = (e) => {
    const toV = this.state.to;
    const fromV = this.state.from;
    const updatedTo = e.target.value;
    if (fromV === updatedTo ) {
      this.setState({
        from: toV,
        to: updatedTo,
        fromValue: '',
        toValue: '',
      })
    }
    else {
      this.setState({
        [e.target.name]: e.target.value,
        fromValue: '',
        toValue: '',
      });
    }
  }

  render() {
    return (
      <div className="App">
        <header className="App-header">
          <h1 className="App-title">Exchange</h1>
        </header>
        <p className="App-intro">
          {currencies.find(c => c.symbol === this.state.from).currency} 
          1 
          = {currencies.find(c => c.symbol === this.state.to).currency} 
          {this.props[this.state.from] && this.props[this.state.from][this.state.to].toFixed(4) }
        </p>
        
        <div>
          <select name='from' value={this.state.from} onChange={this.handleCurrencyFromChange}>
            { currencies.map(currency => 
              <option value={currency.symbol} key={currency.symbol}>
                { currency.symbol }
              </option>)
            }
          </select>
          <input type="number" step="0.01" value={this.state.fromValue} name={this.state.from} onChange={this.handleExchangeRate} />
        </div>
        <div>
          <select name='to' value={this.state.to} onChange={this.handleCurrencyToChange}>
            {currencies.map(currency =>
              <option value={currency.symbol} key={currency.symbol}>
                {currency.symbol}
              </option>)
            }
          </select>
          <input type="number" step="0.01" value={this.state.toValue} name={this.state.to} onChange={this.handleExchangeRate} /> 
        </div>
      </div>
    );
  }
}

const mapStateToProps = ({exchangeRate}) => ({
  ...exchangeRate
});

const mapDispatchToProps = dispatch => ({
  getExchangeRate: (base) => dispatch(getExchangeRate(base))
});

export default connect(mapStateToProps, mapDispatchToProps)(App);
