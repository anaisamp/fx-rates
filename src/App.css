.App {
  text-align: center;
  font-family: 'Roboto', sans-serif;
}

.App-header {
  background-color: #0075eb;
  padding: 20px;
  color: white;
}

.App-title {
  font-size: 1.5em;
}

.App-rates {
  background-color: #0075eb;
  color: white;
  padding: 20px 0 50px;
  margin: 0px;
  font-size: large;
}

.App-convert {
  padding: 50px 0 20px;
}

input, select {
  font-size: 1.2em;
  font-family: 'Roboto', sans-serif;
  margin: 10px 10px;
  padding: 8px 10px;
}
