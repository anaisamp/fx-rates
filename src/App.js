import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import './App.css';

import { getExchangeRate } from './actions/getExchangeRate';
import { updateBalance } from './actions/updateBalance';

import ExchangeRate from './components/exchangeRate';
import ConvertForm from './components/convertForm';

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      from: 'GBP',
      to: 'EUR',
      toValue: '',
      fromValue: '',
    };
  }

  componentDidMount() {
    this.initExchangeRates();
    // this.interval = this.updateExchangeRates(this.state.from, 10000);
  }

  shouldComponentUpdate(nextProps, nextState) {
    return this.props !== nextProps || this.state !== nextState;
  }

  componentWillUnmount() {
    clearInterval(this.interval);
  }

  initExchangeRates = () => {
    this.props.getExchangeRate(this.state.from);
    this.props.getExchangeRate(this.state.to);
  }

  updateExchangeRates = (currency, frequency) => {
    setInterval(() => this.props.getExchangeRate(currency), frequency);
  }

  updateIfNotExists = (updatedCurrency) => {
    if (!this.props[updatedCurrency]) {
      this.props.getExchangeRate(updatedCurrency);
    }
  }

  handleExchangeRate = (e, fromInput, toInput) => {
    const value = e.target.value.replace('-', '');
    const baseCurrency = e.target.name;
    const toCurrency = baseCurrency === this.state.from ? this.state.to : this.state.from;
    const exhangeValue = parseFloat(value) * this.props[baseCurrency][toCurrency];

    this.setState({
      [fromInput]: value,
      [toInput]: exhangeValue === 0 ? exhangeValue.toString() : exhangeValue.toFixed(2).toString(),
    });
  }

  handleCurrencyChange = (e, selectElement) => {
    const fromPreviousCurrency = this.state.from;
    const toPreviousCurrency = this.state.to;
    const nextCurrency = e.target.value;
    const opositySelectElement = selectElement === 'from' ? toPreviousCurrency : fromPreviousCurrency;

    if (opositySelectElement === nextCurrency) {
      this.setState({
        from: selectElement === 'from' ? nextCurrency : toPreviousCurrency,
        to: selectElement === 'from' ? fromPreviousCurrency : nextCurrency,
        fromValue: '',
        toValue: '',
      });
    } else {
      this.setState({
        [e.target.name]: e.target.value,
        fromValue: '',
        toValue: '',
      });
    }

    this.updateIfNotExists(nextCurrency);
  };

  handleExchangeButton = () => {
    const {
      from, fromValue, to, toValue,
    } = this.state;

    if (this.props.balance[from] >= fromValue) {
      const balance = {
        fromValue, from, toValue, to,
      };
      this.props.updateBalance(balance);

      this.setState({
        fromValue: '',
        toValue: '',
      });
    }
  };

  isDisabled = () => !this.state.fromValue || this.props.balance[this.state.from] < this.state.fromValue;

  render() {
    const {
      from, fromValue, to, toValue,
    } = this.state;

    return (
      <div className="App">
        <header>
          <h1 className="App-title">Exchange</h1>
        </header>

        <div className="App-convert">
          <ConvertForm
            selectName="from"
            selectedCurrency={from}
            inputValue={fromValue}
            handleCurrencyChange={e => this.handleCurrencyChange(e, 'from')}
            handleExchangeRate={e => this.handleExchangeRate(e, 'fromValue', 'toValue')}
            balance={this.props.balance[from]}
          />
          <span className="App-rates">
            <ExchangeRate from={from} to={to} rates={this.props} />
          </span>
          <ConvertForm
            selectName="to"
            selectedCurrency={to}
            inputValue={toValue}
            handleCurrencyChange={e => this.handleCurrencyChange(e, 'to')}
            handleExchangeRate={e => this.handleExchangeRate(e, 'toValue', 'fromValue')}
            balance={this.props.balance[to]}
          />
          <button
            className="App-button"
            type="button"
            onClick={this.handleExchangeButton}
            disabled={this.isDisabled()}
          >
              Exchange
          </button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = ({ exchangeRate, balanceUpdate }) => ({
  ...exchangeRate,
  balance: { ...balanceUpdate },
});

const mapDispatchToProps = dispatch => ({
  getExchangeRate: base => dispatch(getExchangeRate(base)),
  updateBalance: data => dispatch(updateBalance(data)),
});

App.propTypes = {
  getExchangeRate: PropTypes.func.isRequired,
  updateBalance: PropTypes.func.isRequired,
  balance: PropTypes.object.isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(App);
