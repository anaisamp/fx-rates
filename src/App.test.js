import React from 'react';
import { shallow, mount } from 'enzyme';
import './config/enzyme-setup';
import { App } from './App';

const props = {
  getExchangeRate: jest.fn(),
  updateBalance: jest.fn(),
  balance: { GBP: '400.30', EUR: '150.00', USD: '21.02' },
  rates: { GBP: { EUR: 1.12, USD: 1.17 }, EUR: { GBP: 0.89, USD: 0.98 } },
};

it('renders App', () => {
  const wrapper = shallow(<App {...props} />);
  expect(wrapper).toMatchSnapshot();
});

it('updates input value "to" when input value "from" has changed', () => {
  const wrapper = mount(<App {...props} />);
  const currency = 'GBP';
  wrapper.find('input').first().simulate('change', { target: { value: '1', name: currency } });
  wrapper.update();
  expect(wrapper.find('input').last().prop('value')).toBe('1.12');
});

it('updates input value "from" when input value "to" has changed', () => {
  const wrapper = mount(<App {...props} />);
  const currency = 'EUR';
  wrapper.find('input').last().simulate('change', { target: { value: '1', name: currency } });
  wrapper.update();
  expect(wrapper.find('input').first().prop('value')).toBe('0.89');
});
