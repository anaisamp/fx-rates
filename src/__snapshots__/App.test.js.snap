// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders App 1`] = `
<div
  className="App"
>
  <header>
    <h1
      className="App-title"
    >
      Exchange
    </h1>
  </header>
  <div
    className="App-convert"
  >
    <ConvertForm
      balance="400.30"
      handleCurrencyChange={[Function]}
      handleExchangeRate={[Function]}
      inputValue=""
      selectName="from"
      selectedCurrency="GBP"
    />
    <span
      className="App-rates"
    >
      <ExchangeRate
        from="GBP"
        rates={
          Object {
            "EUR": Object {
              "GBP": 0.89,
              "USD": 0.98,
            },
            "GBP": Object {
              "EUR": 1.12,
              "USD": 1.17,
            },
          }
        }
        to="EUR"
      />
    </span>
    <ConvertForm
      balance="150.00"
      handleCurrencyChange={[Function]}
      handleExchangeRate={[Function]}
      inputValue=""
      selectName="to"
      selectedCurrency="EUR"
    />
    <button
      className="App-button"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      Exchange
    </button>
  </div>
</div>
`;
